File: ahu_dashboard/lib/models/ahu_state.dart
import 'package:json_annotation/json_annotation.dart';

part 'ahu_state.g.dart';

/// State data from ESP32 AHU unit (retained MQTT message)
@JsonSerializable()
class AhuState {
  final bool run;            // System running state
  final bool m1;             // Motor-1 status
  final bool m2;             // Motor-2 status
  final bool cp;             // Compressor 1 status
  @JsonKey(name: 'cp2') final bool? cp2;  // Compressor 2 status
  @JsonKey(name: 'cpMode') final String? cpMode;  // CP mode: "dual" or "single"
  @JsonKey(name: 'cpActive') final int? cpActive;  // Active CP: 1 or 2
  final bool heater;         // Heater status
  final bool fan;             // Fan status (on/off)
  @JsonKey(name: 'fanSpeed') final int fanSpeed;  // Fan speed: 0=OFF, 1=LOW, 2=MED, 3=HIGH
  final double tempSet;      // Temperature setpoint
  final double humSet;       // Humidity setpoint
  final String ip;           // ESP32 IP address
  @JsonKey(name: 'onlineMode') final bool? onlineMode;  // Operation mode: true=online/cloud, false=offline/local
  
  // Firmware version from ESP32
  final String? version;
  
  // Motor timings (in seconds)
  @JsonKey(name: 'm1_start') final int? m1Start;
  @JsonKey(name: 'm1_post') final int? m1Post;
  @JsonKey(name: 'm2_interval') final int? m2Interval;
  @JsonKey(name: 'm2_run') final int? m2Run;
  @JsonKey(name: 'm2_delay') final int? m2Delay;

  AhuState({
    required this.run,
    required this.m1,
    required this.m2,
    required this.cp,
    this.cp2,
    this.cpMode,
    this.cpActive,
    required this.heater,
    required this.fan,
    required this.fanSpeed,
    required this.tempSet,
    required this.humSet,
    required this.ip,
    this.onlineMode,
    this.version,
    this.m1Start,
    this.m1Post,
    this.m2Interval,
    this.m2Run,
    this.m2Delay,
  });

  factory AhuState.fromJson(Map<String, dynamic> json) =>
      _$AhuStateFromJson(json);

  Map<String, dynamic> toJson() => _$AhuStateToJson(this);
  
  /// Get M2 wait time (interval - run time) for UI display
  /// ESP32 sends actual interval, but UI shows wait time between cycles
  int get m2WaitTime {
    if (m2Interval == null || m2Run == null) return 30;
    // Convert actual interval back to wait time
    return (m2Interval! - m2Run!).clamp(1, 999);
  }

  /// Get fan speed display string
  String get fanSpeedDisplay {
    switch (fanSpeed) {
      case 0:
        return 'OFF';
      case 1:
        return 'LOW (5V)';
      case 2:
        return 'MED (7V)';
      case 3:
        return 'HIGH (9V)';
      default:
        return 'UNKNOWN';
    }
  }
}

File: pi_dashboard/app/models/ahu_state.py
from dataclasses import dataclass
from typing import Optional


@dataclass
class AhuState:
    run: bool = False
    m1: bool = False
    m2: bool = False
    cp: bool = False
    cp2: Optional[bool] = None
    cp_mode: str = "single"        # "dual" | "single"
    cp_active: int = 1             # 1 | 2
    heater: bool = False
    fan: bool = False
    fan_speed: int = 0
    temp_set: float = 24.0
    hum_set: float = 55.0
    ip: str = ""
    online_mode: bool = True
    version: Optional[str] = None
    m1_start: int = 30
    m1_post: int = 10
    m2_interval: int = 3600
    m2_run: int = 30
    m2_delay: int = 5

    @property
    def m2_wait_time(self) -> int:
        return max(0, self.m2_interval - self.m2_run)

    @classmethod
    def from_dict(cls, d: dict) -> "AhuState":
        def g(key, alt=None):
            return d.get(key, alt)

        fan_speed = g("fanSpeed", g("fan_speed", 0))
        try:
            fan_speed = int(fan_speed)
        except (TypeError, ValueError):
            fan_speed = 0

        return cls(
            run=bool(g("run", False)),
            m1=bool(g("m1", False)),
            m2=bool(g("m2", False)),
            cp=bool(g("cp", False)),
            cp2=_optional_bool(g("cp2")),
            cp_mode=g("cpMode", g("cp_mode", "single")) or "single",
            cp_active=int(g("cpActive", g("cp_active", 1)) or 1),
            heater=bool(g("heater", False)),
            fan=bool(g("fan", False)),
            fan_speed=fan_speed,
            temp_set=float(g("tempSet", g("temp_set", 24.0))),
            hum_set=float(g("humSet", g("hum_set", 55.0))),
            ip=str(g("ip", "") or ""),
            online_mode=bool(g("onlineMode", g("online_mode", True))),
            version=g("version"),
            m1_start=int(g("m1Start", g("m1_start", 30)) or 30),
            m1_post=int(g("m1Post", g("m1_post", 10)) or 10),
            m2_interval=int(g("m2Interval", g("m2_interval", 3600)) or 3600),
            m2_run=int(g("m2Run", g("m2_run", 30)) or 30),
            m2_delay=int(g("m2Delay", g("m2_delay", 5)) or 5),
        )


def _optional_bool(v) -> Optional[bool]:
    if v is None:
        return None
    return bool(v)
